File: src/oca/go/src/goca/schemas/virtualnetwork/virtualnetwork.go
/* -------------------------------------------------------------------------- */
/* Copyright 2002-2024, OpenNebula Project, OpenNebula Systems                */
/*                                                                            */
/* Licensed under the Apache License, Version 2.0 (the "License"); you may    */
/* not use this file except in compliance with the License. You may obtain    */
/* a copy of the License at                                                   */
/*                                                                            */
/* http://www.apache.org/licenses/LICENSE-2.0                                 */
/*                                                                            */
/* Unless required by applicable law or agreed to in writing, software        */
/* distributed under the License is distributed on an "AS IS" BASIS,          */
/* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.   */
/* See the License for the specific language governing permissions and        */
/* limitations under the License.                                             */
/*--------------------------------------------------------------------------- */

package virtualnetwork

import (
	"encoding/xml"
	"fmt"

	"github.com/OpenNebula/one/src/oca/go/src/goca/dynamic"
	"github.com/OpenNebula/one/src/oca/go/src/goca/schemas/shared"
)

// Pool represents an OpenNebula VirtualNetwork pool
type Pool struct {
	XMLName         xml.Name         `xml:"VNET_POOL"`
	VirtualNetworks []VirtualNetwork `xml:"VNET"`
}

// VirtualNetwork represents an OpenNebula VirtualNetwork
type VirtualNetwork struct {
	XMLName              xml.Name            `xml:"VNET"`
	ID                   int                 `xml:"ID,omitempty"`
	UID                  int                 `xml:"UID,omitempty"`
	GID                  int                 `xml:"GID,omitempty"`
	UName                string              `xml:"UNAME,omitempty"`
	GName                string              `xml:"GNAME,omitempty"`
	Name                 string              `xml:"NAME"`
	Permissions          *shared.Permissions `xml:"PERMISSIONS,omitempty"`
	Clusters             shared.EntitiesID   `xml:"CLUSTERS,omitempty"`
	Bridge               string              `xml:"BRIDGE,omitempty"`
	BridgeType           string              `xml:"BRIDGE_TYPE,omitempty"` // minOccurs=0
	ParentNetworkID      string              `xml:"PARENT_NETWORK_ID,omitempty"`
	StateRaw             int                 `xml:"STATE,omitempty"`
	VNMad                string              `xml:"VN_MAD"`
	PhyDev               string              `xml:"PHYDEV,omitempty"`
	VlanID               string              `xml:"VLAN_ID,omitempty"`       // minOccurs=0
	OuterVlanID          string              `xml:"OUTER_VLAN_ID,omitempty"` // minOccurs=0
	VlanIDAutomatic      string              `xml:"VLAN_ID_AUTOMATIC,omitempty,omitempty"`
	OuterVlanIDAutomatic string              `xml:"OUTER_VLAN_ID_AUTOMATIC,omitempty"`
	UsedLeases           int                 `xml:"USED_LEASES,omitempty"`
	VRouters             shared.EntitiesID   `xml:"VROUTERS,omitempty"`
	Template             Template            `xml:"TEMPLATE"`

	// Variable parts between one.vnpool.info and one.vn.info
	ARs  []AR         `xml:"AR_POOL>AR,omitempty"`
	Lock *shared.Lock `xml:"LOCK,omitempty"`
}

type AR struct {
	ID                string        `xml:"AR_ID,omitempty"`
	GlobalPrefix      string        `xml:"GLOBAL_PREFIX,omitempty"` // minOccurs=0
	IP                string        `xml:"IP"`                      // minOccurs=0
	MAC               string        `xml:"MAC,omitempty"`
	ParentNetworkARID string        `xml:"PARENT_NETWORK_AR_ID,omitempty"` // minOccurs=0
	Size              int           `xml:"SIZE"`
	Type              string        `xml:"TYPE"`
	ULAPrefix         string        `xml:"ULA_PREFIX,omitempty"` // minOccurs=0
	VNMAD             string        `xml:"VN_MAD,omitempty"`     // minOccurs=0
	MACEnd            string        `xml:"MAC_END,omitempty"`
	IPEnd             string        `xml:"IP_END,omitempty"`
	IP6ULA            string        `xml:"IP6_ULA,omitempty"`
	IP6ULAEnd         string        `xml:"IP6_ULA_END,omitempty"`
	IP6Global         string        `xml:"IP6_GLOBAL,omitempty"`
	IP6GlobalEnd      string        `xml:"IP6_GLOBAL_END,omitempty"`
	IP6               string        `xml:"IP6,omitempty"`
	IP6End            string        `xml:"IP6_END,omitempty"`
	UsedLeases        string        `xml:"USED_LEASES,omitempty"`
	Leases            []Lease       `xml:"LEASES>LEASE,omitempty"`
	Custom            dynamic.Pairs `xml:",any,omitempty"`

	// Not filled with Info
	Allocated string `xml:"ALLOCATED"`
}

type Lease struct {
	IP        string `xml:"IP,omitempty"`
	IP6       string `xml:"IP6,omitempty"`
	IP6Global string `xml:"IP6GLOBAL,omitempty"`
	IP6Link   string `xml:"IP6LINK,omitempty"`
	IP6ULA    string `xml:"IP6ULA,omitempty"`
	MAC       string `xml:"MAC,omitempty"`
	VM        int    `xml:"VM,omitempty"`
	VNet      int    `xml:"VNET,omitempty"`
	VRouter   int    `xml:"VROUTER,omitempty"`
}

// State is the state of the Virtual Network
type State int

const (
	// Init Virtual Network is being initialized
	Init State = iota

	// Ready Virtual Network is ready to be used
	Ready

	// LockCreate Virtual Network driver vnet_create pending
	LockCreate

	// LockDelete Virtual Network driver vnet_delete pending
	LockDelete

	// Done Virtual Netowrk finalized
	Done

	// Error Virtual Network is in error state
	Error
)

func (s State) isValid() bool {
	if s >= Init && s <= Error {
		return true
	}
	return false
}

// String returns the string version of the State
func (s State) String() string {
	return [...]string{
		"INIT",
		"READY",
		"LOCK_CREATE",
		"LOCK_DELETE",
		"DONE",
		"ERROR",
	}[s]
}

// State looks up the state of the Virtual Network and returns the State
func (vn *VirtualNetwork) State() (State, error) {
	state := State(vn.StateRaw)
	if !state.isValid() {
		return -1, fmt.Errorf("Virtual Network State: this state value is not currently handled: %d\n", vn.StateRaw)
	}
	return state, nil
}

// StateString returns the state in string format
func (vn *VirtualNetwork) StateString() (string, error) {
	state := State(vn.StateRaw)
	if !state.isValid() {
		return "", fmt.Errorf("Virtual Network State: this state value is not currently handled: %d\n", vn.StateRaw)
	}
	return state.String(), nil
}

File: src/fireedge/src/client/models/VirtualNetwork.js
/* ------------------------------------------------------------------------- *
 * Copyright 2002-2024, OpenNebula Project, OpenNebula Systems               *
 *                                                                           *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may   *
 * not use this file except in compliance with the License. You may obtain   *
 * a copy of the License at                                                  *
 *                                                                           *
 * http://www.apache.org/licenses/LICENSE-2.0                                *
 *                                                                           *
 * Unless required by applicable law or agreed to in writing, software       *
 * distributed under the License is distributed on an "AS IS" BASIS,         *
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  *
 * See the License for the specific language governing permissions and       *
 * limitations under the License.                                            *
 * ------------------------------------------------------------------------- */
import { isIPv6, isIPv4, isMAC } from 'client/utils'
import {
  VirtualNetwork,
  AddressRange,
  VN_STATES,
  STATES,
  LEASE_STATE,
  VN_ACTIONS_BY_STATE,
} from 'client/constants'

/**
 * Returns the lease state.
 *
 * @param {string} state - Leases
 * @returns {STATES.StateInfo} State information from leases
 */
export const getLeaseState = (state) => LEASE_STATE[state]

/**
 * Returns the state of the virtual network.
 *
 * @param {VirtualNetwork} virtualNetwork - Virtual network
 * @returns {STATES.StateInfo} State information from resource
 */
export const getState = ({ STATE = 0 } = {}) => VN_STATES[+STATE]

/**
 * Returns the Virtual Network Manager name.
 *
 * @param {VirtualNetwork} virtualNetwork - Virtual network
 * @returns {string} Virtual Network Manager
 */
export const getVNManager = (virtualNetwork) => virtualNetwork?.VN_MAD

/**
 * Returns the total number of leases in the virtual network.
 *
 * @param {VirtualNetwork} virtualNetwork - Virtual network
 * @returns {number} Total leases
 */
export const getTotalLeases = ({ AR_POOL } = {}) => {
  const addressRanges = [AR_POOL?.AR ?? []].flat()

  return addressRanges.reduce((total, { SIZE = 0 }) => total + +SIZE, 0)
}

/**
 * Returns the virtual network leases information.
 *
 * @param {VirtualNetwork} virtualNetwork - Virtual network
 * @returns {{ percentOfUsed: number, percentLabel: string }} Leases information
 */
export const getLeasesInfo = ({ USED_LEASES, ...virtualNetwork } = {}) => {
  const totalLeases = getTotalLeases(virtualNetwork)
  const percentOfUsed = (+USED_LEASES * 100) / +totalLeases || 0
  const percentLabel = `${USED_LEASES} / ${totalLeases} (${Math.round(
    percentOfUsed
  )}%)`

  return { percentOfUsed, percentLabel }
}

/**
 * Returns the address range leases information.
 *
 * @param {AddressRange} ar - Address range
 * @returns {{ percentOfUsed: number, percentLabel: string }} Leases information
 */
export const getARLeasesInfo = ({ USED_LEASES, SIZE } = {}) => {
  const percentOfUsed = (+USED_LEASES * 100) / +SIZE || 0
  const percentLabel = `${USED_LEASES} / ${SIZE} (${Math.round(
    percentOfUsed
  )}%)`

  return { percentOfUsed, percentLabel }
}

/**
 * Checks the address type: IP, IP6 or MAC
 * Otherwise returns undefined.
 *
 * @param {string} addr - Address to check
 * @returns {'IP'|'IP6'|'MAC'|undefined} Returns name of address type, undefined otherwise
 */
export const getAddressType = (addr) => {
  if (isIPv4(addr)) return 'IP'
  if (isIPv6(addr)) return 'IP6'
  if (isMAC(addr)) return 'MAC'
}

/**
 * Check if action is available for **all Virtual Networks**.
 *
 * @param {object} action - Virtual Network action
 * @param {VirtualNetwork|VirtualNetwork[]} vnets - Virtual networks
 * @returns {boolean} If `true`, the action is available for all Virtual Networks
 */
export const isAvailableAction = (action, vnets = []) => {
  if (VN_ACTIONS_BY_STATE[action]?.length === 0) return true

  return [vnets].flat().every((vnet) => {
    const state = VN_STATES[vnet.STATE]?.name

    return VN_ACTIONS_BY_STATE[action]?.includes(state)
  })
}
